### DISCUSSION TITLE
Adding printing to `solgo.ast`

### DISCUSSION THREAD

**Discussion by xianlinc** [unpackdev/solgo#188] [Proposals]
Hey @0x19, 

Big thanks for `solgo`! My teammate and I are hacking away on a Solidity to Solidity optimizing compiler, and your project is a lifesaver. 

We went ahead and [forked](https://github.com/yongkangc/solgo) your repo and we added a new method to all AST nodes for recursive source code printing, but it's still very much in the tinkering stage   (think prototype with bugs).

Just wanted to see if you'd be curious about our progress once we have a more solid prototype. Maybe then we can chat about potentially integrating it into `solgo`, if that aligns with your vision for the project.

Thanks again for the awesome work!

**Comment by 0x19** [comment]
Hey @xianlinc, first thank you! I wasn't sure if anyone else would use this project as in general, Golang is not the best for this type of workload (missing/weak pattern/sum matching). It can be done as seen but it was a tremendous pain in the rear-end to get it done. 

The thing you are doing down there is a huge task that I have in list to do but didn't just had time as it would allow to do exactly what you are doing, among other things. 

I am here to help out. One thing I'd suggest please don't build strings like you're doing it now. `ToSource()` is great, however, please use strings.Builder as it's much faster then what you're doing at this moment. You can see example (not-related but example never the less) at: https://github.com/unpackdev/solgo/blob/main/cfg/builder_printer.go#L30

Longer story short, I am very much interested and would love to integrate.  I'd as well suggest for you to use latest master branch as I've applied serious amount of fixes into the ast library in the past couple of days. For example, array type names were broken to a point. 

Would as well later on love to add your project to projects in the readme if you guys are ok with it!

**Reply by xianlinc** [reply]
Thanks for the fixes and the suggestions! We will for sure do a round of refactoring before we make any PRs to your library 😄. Right now we are just building as fast as possible to get a prototype up. 

Also, we would love to have our project in the readme :)

**Reply by 0x19** [reply]
Sounds good!

Can you please open up a new PR and add it to this section :)

https://github.com/unpackdev/solgo?tab=readme-ov-file#external-projects--extensions

I will gladly approve the PR :) 

p.s. will be the first one besides me!

**Comment by 0x19** [comment]
Thinking potentially about some sort of generic printer interface. Not sure if that would be a overkill.

**Comment by 0x19** [comment]
As well if you wish to join, we can as well discuss on Discord. Here's the invite -> https://discord.com/invite/PdHVbuTQRX

**Comment by xianlinc** [comment]
relevant PR #201
